pgmax/fg/nodes.py

```diff
@@ -174,3 +174,120 @@ def compile_wiring(
             var_states_for_edges=var_states_for_edges,
             factor_configs_edge_states=self.factor_configs_edge_states,
         )
+
+
+@jax.tree_util.register_pytree_node_class
+@dataclass(frozen=True, eq=False)
+class ORWiring:
+    """Wiring for OR factors.
+
+    Args:
+        parents_var_states_for_edges: Array of shape (2 * num_parents,)
+            Global parent variable state indices for each edge state
+        child_var_states_for_edges: Array of shape (2,)
+            Global child variable state indices for each edge state
+        parents_states: Array of shape (num_parents, 2)
+            parents_states[ii, 0] contains the global factor index
+            parents_states[ii, 1] contains the message index of the parent variable's state 0.
+            The parent variable's state 1 is parents_states[ii, 2] + 1
+        children_states: Array of shape (num_factors,)
+            children_states[ii] contains the message index of the child variable's state 0
+            The child variable's state 1 is children_states[ii, 1] + 1
+    """
+
+    # TODO: is it needed?
+    parents_var_states_for_edges: Union[np.ndarray, jnp.ndarray]
+    child_var_states_for_edges: Union[np.ndarray, jnp.ndarray]
+    parents_states: Union[np.ndarray, jnp.ndarray]
+    children_states: Union[np.ndarray, jnp.ndarray]
+
+    def __post_init__(self):
+        for field in self.__dataclass_fields__:
+            if isinstance(getattr(self, field), np.ndarray):
+                getattr(self, field).flags.writeable = False
+
+    def tree_flatten(self):
+        return jax.tree_util.tree_flatten(asdict(self))
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        return cls(**aux_data.unflatten(children))
+
+
+@dataclass(frozen=True, eq=False)
+class ORFactor:
+    """An OR factor
+
+    Args:
+        parents_variables: List of parents variables
+        child_variable: Child variable
+
+    Raises:
+        ValueError: If:
+            (1) The parents variables are not all binary
+            (2) The child variable is not binary
+            (3) The list of parent variables is empty
+    """
+
+    parents_variables: Tuple[Variable, ...]
+    child_variable: Variable
+    log_potentials: None
```

Comment:
This looks redundant (since there is not need to be consistent with anything here)...

Reply:
this is needed for the `factor_group_log_potentials` method of the `FactorGroup` class